lib/pedicel/base.rb

```diff
@@ -140,20 +166,27 @@ def verify_signature(ca_certificate_pem: Pedicel.config[:apple_root_ca_g3_cert_p
       self
     end
 
-    private
+    def self.verify_signature_certificate_oids(signature:, config: Pedicel.config)
+      leafs, intermediates = [], []
+
+      signature.certificates.each do |certificate|
+        certificate.extensions.each do |extension|
+          case extension.oid
+          when config[:oids][:leaf_certificate]         then leafs << certificate
+          when config[:oids][:intermediate_certificate] then intermediates << certificate
+          end
+        end
+      end
```

Comment:
This won't work:

`signature.certificates` is:

```
=> [#<OpenSSL::X509::Certificate
  subject=#<OpenSSL::X509::Name CN=ecc-smp-broker-sign_UC4-PROD,OU=pOS Systems,O=Pedicel Inc.,C=DK>,
  issuer=#<OpenSSL::X509::Name CN=Pedicel Application Integration CA - G3,OU=Pedicel Certification Authority,O=Pedicel Inc.,C=DK>,
  serial=#<OpenSSL::BN 1>,
  not_before=2017-01-01 00:00:00 UTC,
  not_after=2020-01-01 00:00:00 UTC>,
 #<OpenSSL::X509::Certificate
  subject=#<OpenSSL::X509::Name CN=Pedicel Application Integration CA - G3,OU=Pedicel Certification Authority,O=Pedicel Inc.,C=DK>,
  issuer=#<OpenSSL::X509::Name CN=Pedicel Root CA - G3,OU=Pedicel Certification Authority,O=Pedicel Inc.,C=DK>,
  serial=#<OpenSSL::BN 1>,
  not_before=2017-01-01 00:00:00 UTC,
  not_after=2020-01-01 00:00:00 UTC>,
 #<OpenSSL::X509::Certificate
  subject=#<OpenSSL::X509::Name CN=Pedicel Root CA - G3,OU=Pedicel Certification Authority,O=Pedicel Inc.,C=DK>,
  issuer=#<OpenSSL::X509::Name CN=Pedicel Root CA - G3,OU=Pedicel Certification Authority,O=Pedicel Inc.,C=DK>,
  serial=#<OpenSSL::BN 1>,
  not_before=2017-01-01 00:00:00 UTC,
  not_after=2020-01-01 00:00:00 UTC>]
```

However, `signature.signers` is:

```
=> [#<OpenSSL::PKCS7::SignerInfo:0x00000000016e6210>, #<OpenSSL::PKCS7::SignerInfo:0x00000000016e61e8>, #<OpenSSL::PKCS7::SignerInfo:0x00000000016e6170>]
```

But `signature.verify(...)` is still happy although there are multiple signers.

Notice that `.verify_signed_time` already has this uniqueness check. Then I wonder if `.verify_signed_time` should in fact be more forgiving. We could accept `leaf` as a parameter to `.verify_signed_time` and accept the signing time if `>= 1` of the signers is the leaf and the `signing_time` is "good". @kse-clearhaus WDYT?

Reply:
As I interpret the question, you are saying that there are 3 signers on the signature?
I would guess that there should only be 1 and that `pedicel-pay` should not sign with intermediate/root certificates.

However, the [specification 1.c](https://developer.apple.com/library/content/documentation/PassKit/Reference/PaymentTokenJSON/PaymentTokenJSON.html) requires that  we:

> ... ensure that the signature was created using the private key corresponding to the leaf certificate

I don't see a problem in accepting multiple signatures, however I wouldn't expect them to appear in production.

Giving the code a closer look, my tests are currently failing.